_modules/ext_module_qvm.py

```diff
@@ -941,8 +941,197 @@ def prefs(vmname, *varargs, **kwargs):
+              - pci:dom0:01_00.0: []
+              - pci:dom0:02_00.0:
+                - pci_strictreset: false
+              - bridge:sys-net-interfaces:br0:
+                - ip: 192.168.0.1
+                - netmask: 255.255.255.0
+            - detach:
+              - pci:dom0:28_00.3: []
+              - bridge:sys-net-interfaces:br1: []
+    """
+
+    # CLI 'qubesctl qvm.devices <vm_name> (attach|detach) device [device...]'
+    if varargs and varargs[0] in ['attach', 'detach']:
+        devices = []
+        for device in varargs[1:]:
+            device_type, backend, dev_id = device.split(':')
+            devices.append({'device_type': device_type, 'backend': backend, 'dev_id': dev_id})
+        if devices:
+            kwargs[varargs[0]] = devices
+
+    # Convert kwargs to string for argparse
+    for action in ['attach', 'detach']:
+        if action in kwargs:
+            kwargs[action] = json.dumps(kwargs[action])
+
+    # Set default status-mode to show all status entries
+    kwargs.setdefault('status-mode', 'all')
+
+    qvm = _QVMBase('qvm.devices', **kwargs)
+    qvm.parser.add_argument(
+        'vmname',
+        action=_VMAction,
+        help='Virtual machine name'
+    )
+    qvm.parser.add_argument('--list', nargs='*', help='List devices')
+    qvm.parser.add_argument(
+        '--attach',
+        nargs='*',
+        default=[],
+        help='List of devices to attach'
+    )
+    qvm.parser.add_argument(
+        '--detach',
+        nargs='*',
+        default=[],
+        help='List of devices to detach'
+    )
+
+    args = qvm.parse_args(vmname, *varargs, **kwargs)
+
+    current_devices = []
+    for device_type in ['pci', 'bridge']:
+        for device in args.vm.devices[device_type].assignments():
+            current_devices.append(
+                {'device_type': device_type, 'backend': device.backend_domain.name, 'dev_id': device.ident,
+                 'options': device.options})
+
+    # Return all current devices if a 'list' only was selected
+    if args.list is not None or not (args.attach or args.detach):
+        message = []
+        for device in current_devices:
+            msg_options = '(' + ', '.join('{}={}'.format(key, value) for key, value in device['options'].items()) + ')'
+            message.append(
+                '    ' + device['device_type'] + ':' + device['backend'] + ':' + device['dev_id'] + ' ' + msg_options)
+
+        return {'result': True, 'comment': '[ATTACHED]:\n' + '\n'.join(message)}
+
+    # Helping function for parsing devices in args.attach and args.detach
+    def parse_device(raw_dev):
+        """
+        Example:
+            args.attach = [{u'bridge:sys-bridge:bridge0': [{u'ip': u'192.168.0.1'}, {u'netmask': u'255.255.255.0'}]}]
+        """
+        inline_dev = list(raw_dev.keys())[0]
+        device = {'device_type': inline_dev.split(':')[0], 'backend': inline_dev.split(':')[1],
+                  'dev_id': inline_dev.split(':')[2], 'options': {}}
```

Comment:
Some meaningful error message about not enough `:`?

Reply:
Sure.